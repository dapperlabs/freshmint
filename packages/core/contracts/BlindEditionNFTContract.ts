// @ts-ignore
import * as fcl from '@onflow/fcl';
// @ts-ignore
import * as t from '@onflow/types';

import { NFTContract, CollectionMetadata, prepareCollectionMetadata, Royalty, prepareRoyalties } from './NFTContract';
import { MetadataMap } from '../metadata';
import { BlindEditionNFTGenerator } from '../generators/BlindEditionNFTGenerator';
import { hashValuesWithSalt } from '../hash';
import { FreshmintConfig, ContractImports } from '../config';
import { Transaction, TransactionResult } from '../transactions';
import { PublicKey, SignatureAlgorithm, HashAlgorithm } from '../crypto';

export type EditionInput = {
  size: number;
  metadata: MetadataMap;
};

export type EditionNFT = {
  editionId: string;
  editionSerial: string;
};

export type EditionResult = {
  id: string;
  metadata: MetadataMap;
  size: number;
  nfts: EditionNFT[];
};

export type HashedEditionNFT = {
  editionId: string;
  editionSerial: string;
  editionHash: string;
  editionSalt: string;
};

export type NFTMintResult = {
  id: string;
  editionId: string;
  editionSerial: string;
  editionHash: string;
  editionSalt: string;
  transactionId: string;
};

export interface NFTRevealInput {
  id: string;
  editionId: string;
  editionSerial: string;
  editionSalt: string;
}

export type NFTRevealResult = {
  id: string;
  transactionId: string;
};

export class BlindEditionNFTContract extends NFTContract {
  getSource(imports: ContractImports, options?: { saveAdminResourceToContractAccount?: boolean }): string {
    return BlindEditionNFTGenerator.contract({
      imports,
      contractName: this.name,
      schema: this.schema,
      saveAdminResourceToContractAccount: options?.saveAdminResourceToContractAccount,
    });
  }

  deploy({
    publicKey,
    hashAlgorithm,
    placeholderImage,
    collectionMetadata,
    royalties,
    saveAdminResourceToContractAccount,
  }: {
    publicKey: PublicKey;
    hashAlgorithm: HashAlgorithm;
    placeholderImage: string;
    collectionMetadata: CollectionMetadata;
    royalties?: Royalty[];
    saveAdminResourceToContractAccount?: boolean;
  }): Transaction<string> {
    return new Transaction(
      ({ imports }: FreshmintConfig) => {
        const script = BlindEditionNFTGenerator.deployToNewAccount({ imports });

        const contractCode = this.getSource(imports, { saveAdminResourceToContractAccount });
        const contractCodeHex = Buffer.from(contractCode, 'utf-8').toString('hex');

        const sigAlgo = publicKey.signatureAlgorithm();

        const { royaltyAddresses, royaltyReceiverPaths, royaltyCuts, royaltyDescriptions } = prepareRoyalties(
          royalties ?? [],
        );

        return {
          script,
          args: [
            fcl.arg(this.name, t.String),
            fcl.arg(contractCodeHex, t.String),
            fcl.arg(publicKey.toHex(), t.String),
            fcl.arg(SignatureAlgorithm.toCadence(sigAlgo), t.UInt8),
            fcl.arg(HashAlgorithm.toCadence(hashAlgorithm), t.UInt8),
            fcl.arg(placeholderImage, t.String),
            fcl.arg(prepareCollectionMetadata(imports.MetadataViews, collectionMetadata), t.Identity),
            fcl.arg(royaltyAddresses, t.Array(t.Address)),
            fcl.arg(royaltyReceiverPaths, t.Array(t.Path)),
            fcl.arg(royaltyCuts, t.Array(t.UFix64)),
            fcl.arg(royaltyDescriptions, t.Array(t.String)),
            fcl.arg(saveAdminResourceToContractAccount ?? false, t.Bool),
          ],
          computeLimit: 1000,
          signers: this.getSigners(),
        };
      },
      ({ events }: TransactionResult) => {
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        const accountCreatedEvent = events.find((event) => event.type === 'flow.AccountCreated')!;

        const address = accountCreatedEvent.data['address'];

        this.setAddress(address);

        return address;
      },
    );
  }

  createEdition(edition: EditionInput): Transaction<EditionResult> {
    const editions = [edition];

    return new Transaction(
      this.makeCreateEditionsTransaction(editions),
      (result: TransactionResult) => formatEditionResults(result, editions)[0],
    );
  }

  createEditions(editions: EditionInput[]): Transaction<EditionResult[]> {
    return new Transaction(this.makeCreateEditionsTransaction(editions), (result: TransactionResult) =>
      formatEditionResults(result, editions),
    );
  }

  private makeCreateEditionsTransaction(editions: EditionInput[]) {
    return ({ imports }: FreshmintConfig) => {
      const script = BlindEditionNFTGenerator.createEditions({
        imports,
        contractName: this.name,
        contractAddress: this.getAddress(),
        schema: this.schema,
      });

      const sizes = editions.map((edition) => edition.size.toString(10));

      return {
        script,
        args: [
          fcl.arg(sizes, t.Array(t.UInt)),
          ...this.schema.fields.map((field) => {
            return fcl.arg(
              editions.map((edition) => field.getValue(edition.metadata)),
              t.Array(field.asCadenceTypeObject()),
            );
          }),
        ],
        computeLimit: 9999,
        signers: this.getSigners(),
      };
    };
  }

  mintNFTs(nfts: EditionNFT[], { bucket }: { bucket?: string } = {}): Transaction<NFTMintResult[]> {
    const hashedNFTs = this.hashNFTs(nfts);
    return this.mintHashedNFTs(hashedNFTs, { bucket });
  }

  mintHashedNFTs(hashedNFTs: HashedEditionNFT[], { bucket }: { bucket?: string } = {}): Transaction<NFTMintResult[]> {
    return new Transaction(
      ({ imports }: FreshmintConfig) => {
        const script = BlindEditionNFTGenerator.mint({
          imports,
          contractName: this.name,
          contractAddress: this.getAddress(),
        });

        const hashes = hashedNFTs.map((nft) => nft.editionHash);

        return {
          script,
          args: [fcl.arg(hashes, t.Array(t.String)), fcl.arg(bucket, t.Optional(t.String))],
          computeLimit: 9999,
          signers: this.getSigners(),
        };
      },
      (result) => formatMintResults(result, hashedNFTs),
    );
  }

  hashNFTs(nfts: EditionNFT[]): HashedEditionNFT[] {
    return nfts.map((nft) => {
      const { hash, salt } = hashEdition(nft.editionId, nft.editionSerial);

      return {
        editionId: nft.editionId,
        editionSerial: nft.editionSerial,
        editionHash: hash.toString('hex'),
        editionSalt: salt.toString('hex'),
      };
    });
  }

  revealNFT(nft: NFTRevealInput): Transaction<NFTRevealResult> {
    const nfts = [nft];

    return new Transaction(
      this.makeRevealNFTsTransaction(nfts),
      (result: TransactionResult) => formatRevealResults(result)[0],
    );
  }

  revealNFTs(nfts: NFTRevealInput[]): Transaction<NFTRevealResult[]> {
    return new Transaction(this.makeRevealNFTsTransaction(nfts), (result: TransactionResult) =>
      formatRevealResults(result),
    );
  }

  private makeRevealNFTsTransaction(nfts: NFTRevealInput[]) {
    return ({ imports }: FreshmintConfig) => {
      const script = BlindEditionNFTGenerator.reveal({
        imports,
        contractName: this.name,
        // TODO: return error if contract address is not set
        contractAddress: this.address ?? '',
      });

      const nftIds = nfts.map((nft) => nft.id);
      const editionIds = nfts.map((nft) => nft.editionId);
      const serialNumbers = nfts.map((nft) => nft.editionSerial);
      const salts = nfts.map((nft) => nft.editionSalt);

      return {
        script,
        args: [
          fcl.arg(nftIds, t.Array(t.UInt64)),
          fcl.arg(editionIds, t.Array(t.UInt64)),
          fcl.arg(serialNumbers, t.Array(t.UInt64)),
          fcl.arg(salts, t.Array(t.String)),
        ],
        computeLimit: 9999,
        signers: this.getSigners(),
      };
    };
  }
}

function formatMintResults({ transactionId, events }: TransactionResult, nfts: HashedEditionNFT[]): NFTMintResult[] {
  const deposits = events.filter((event) => event.type.includes('.Minted'));

  return deposits.map((deposit, i) => {
    const { editionId, editionSerial, editionHash, editionSalt } = nfts[i];

    return {
      id: deposit.data.id,
      editionId,
      editionSerial,
      editionHash,
      editionSalt,
      transactionId,
    };
  });
}

function formatEditionResults({ events }: TransactionResult, editions: EditionInput[]): EditionResult[] {
  const editionEvents = events.filter((event) => event.type.includes('.EditionCreated'));

  return editions.flatMap((edition, i) => {
    const editionEvent: any = editionEvents[i];
    const editionId = editionEvent.data.edition.id;

    return {
      id: editionId,
      metadata: edition.metadata,
      size: edition.size,
      nfts: getEditionNFTs(editionId, edition.size),
    };
  });
}

function formatRevealResults({ transactionId, events }: TransactionResult): NFTRevealResult[] {
  const deposits = events.filter((event) => event.type.includes('.Revealed'));

  return deposits.map((deposit) => {
    return {
      id: deposit.data.id,
      transactionId,
    };
  });
}

function getEditionNFTs(id: string, size: number): EditionNFT[] {
  const nfts = Array(size);

  for (let i = 0; i < size; i++) {
    nfts[i] = {
      editionId: id,
      editionSerial: String(i + 1),
    };
  }

  return nfts;
}

function hashEdition(editionId: string, editionSerial: string) {
  // TODO: use big-endian bytes
  const editionIdBuffer = Buffer.from(editionId, 'utf-8');
  const editionSerialBuffer = Buffer.from(editionSerial, 'utf-8');

  return hashValuesWithSalt([editionIdBuffer, editionSerialBuffer]);
}
