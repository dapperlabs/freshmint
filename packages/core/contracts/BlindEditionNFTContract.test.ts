import { BlindEditionNFTContract, EditionResult, NFTMintResult } from './BlindEditionNFTContract';
import { FreshmintClaimSaleContract } from './FreshmintClaimSaleContract';

import {
  client,
  contractHashAlgorithm,
  contractPublicKey,
  ownerAuthorizer,
  getTestSchema,
  setupEmulator,
  teardownEmulator,
  collectionMetadata,
  placeholderImage,
} from '../testHelpers';

describe('BlindEditionNFTContract', () => {
  beforeAll(setupEmulator);
  afterAll(teardownEmulator);

  const contract = new BlindEditionNFTContract({
    name: 'BlindEditionNFT_Test',
    schema: getTestSchema(false),
    owner: ownerAuthorizer,
  });

  it('should generate a contract', async () => {
    expect(contract.getSource(client.config.imports)).toMatchSnapshot();
  });

  it('should deploy a contract', async () => {
    await client.send(
      contract.deploy({
        publicKey: contractPublicKey,
        hashAlgorithm: contractHashAlgorithm,
        placeholderImage,
        collectionMetadata,
      }),
    );
  });

  // TODO: refactor this test case into a separate test suite
  it('should deploy a contract and save the admin resource to the contract account', async () => {
    const contractWithAdmin = new BlindEditionNFTContract({
      name: 'BlindEditionNFT_Test_ContractAdmin',
      schema: getTestSchema(false),
      owner: ownerAuthorizer,
    });

    await client.send(
      contractWithAdmin.deploy({
        publicKey: contractPublicKey,
        hashAlgorithm: contractHashAlgorithm,
        placeholderImage: 'sample-image.jpeg',
        collectionMetadata,
        saveAdminResourceToContractAccount: true,
      }),
    );
  });

  let edition1: EditionResult;
  let edition2: EditionResult;

  it('should create edition 1', async () => {
    const editionMetadata1 = {
      size: 5,
      metadata: {
        name: 'Edition 1',
        description: 'This is the first edition.',
        thumbnail: 'edition-1.jpeg',
      },
    };

    edition1 = await client.send(contract.createEdition(editionMetadata1));
  });

  it('should create edition 2', async () => {
    const editionMetadata2 = {
      size: 5,
      metadata: {
        name: 'Edition 2',
        description: 'This is the second edition.',
        thumbnail: 'edition-2.jpeg',
      },
    };

    edition2 = await client.send(contract.createEdition(editionMetadata2));
  });

  let edition1mintedNFTs: NFTMintResult[];
  let edition2mintedNFTs: NFTMintResult[];

  it('should mint Edition 1 NFTs into default bucket', async () => {
    edition1mintedNFTs = await client.send(contract.mintNFTs(edition1.nfts));
  });

  const edition2Bucket = 'edition2';

  it('should mint Edition 2 NFTs into custom bucket', async () => {
    edition2mintedNFTs = await client.send(contract.mintNFTs(edition2.nfts, { bucket: edition2Bucket }));
  });

  const sale = new FreshmintClaimSaleContract(contract);

  const sale1 = 'sale1';
  const sale2 = 'sale2';

  describe('sale 1', () => {
    it('should start a sale from default bucket', async () => {
      await client.send(sale.start({ id: sale1, price: '10.0' }));
    });

    it('should claim an NFT', async () => {
      await client.send(sale.claimNFT(ownerAuthorizer.address, ownerAuthorizer, sale1));
    });

    it('should stop a sale', async () => {
      await client.send(sale.stop(sale1));
    });
  });

  describe('sale 2', () => {
    it('should start a sale from custom bucket', async () => {
      await client.send(sale.start({ id: sale2, price: '10.0', bucket: edition2Bucket }));
    });

    it('should claim an NFT', async () => {
      await client.send(sale.claimNFT(ownerAuthorizer.address, ownerAuthorizer, sale2));
    });

    it('should stop a sale', async () => {
      await client.send(sale.stop(sale2));
    });
  });

  it('should reveal edition 1 NFTs', async () => {
    await client.send(contract.revealNFTs(edition1mintedNFTs));
  });

  it('should reveal edition 2 NFTs', async () => {
    await client.send(contract.revealNFTs(edition2mintedNFTs));
  });
});
