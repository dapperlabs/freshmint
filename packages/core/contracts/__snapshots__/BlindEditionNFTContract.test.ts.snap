// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlindEditionNFTContract should generate a contract 1`] = `
"import NonFungibleToken from 0xf8d6e0586b0a20c7
import MetadataViews from 0xf8d6e0586b0a20c7
import FungibleToken from 0xee82856bf20e2aa6
import FreshmintMetadataViews from 0xf8d6e0586b0a20c7

pub contract BlindEditionNFT_Test: NonFungibleToken {

    pub let version: String

    pub event ContractInitialized()
    pub event Withdraw(id: UInt64, from: Address?)
    pub event Deposit(id: UInt64, to: Address?)
    pub event Minted(id: UInt64, editionHash: [UInt8])
    pub event Revealed(id: UInt64, editionID: UInt64, editionSerial: UInt64)
    pub event Burned(id: UInt64)
    pub event EditionCreated(edition: Edition)

    pub let CollectionStoragePath: StoragePath
    pub let CollectionPublicPath: PublicPath
    pub let CollectionPrivatePath: PrivatePath
    pub let AdminStoragePath: StoragePath

    /// The total number of BlindEditionNFT_Test NFTs that have been minted.
    ///
    pub var totalSupply: UInt64

    /// The total number of BlindEditionNFT_Test editions that have been created.
    ///
    pub var totalEditions: UInt64

    /// A placeholder image used to display NFTs that have not yet been revealed.
    pub let placeholderImage: String

    /// A list of royalty recipients that is attached to all NFTs
    /// minted by this contract.
    ///
    access(contract) var royalties: [MetadataViews.Royalty]
    
    /// Return the royalty recipients for this contract.
    ///
    pub fun getRoyalties(): [MetadataViews.Royalty] {
        return BlindEditionNFT_Test.royalties
    }

    pub struct Metadata {
    
        pub let name: String
        pub let description: String
        pub let thumbnail: String

        init(
            name: String,
            description: String,
            thumbnail: String,
        ) {
            self.name = name
            self.description = description
            self.thumbnail = thumbnail
        }
    }

    pub struct Edition {

        pub let id: UInt64
        pub let size: UInt
        pub let metadata: Metadata

        init(
            id: UInt64,
            size: UInt,
            metadata: Metadata
        ) {
            self.id = id
            self.size = size
            self.metadata = metadata
        }
    }

    access(self) let editions: {UInt64: Edition}

    pub fun getEdition(id: UInt64): Edition? {
        return BlindEditionNFT_Test.editions[id]
    }

    pub struct EditionMember {
    
        pub let editionID: UInt64
        pub let editionSerial: UInt64
        pub let editionSalt: [UInt8]

        init(
            editionID: UInt64,
            editionSerial: UInt64,
            editionSalt: [UInt8],
        ) {
            self.editionID = editionID
            self.editionSerial = editionSerial
            self.editionSalt = editionSalt
        }

        pub fun getData(): Edition {
            return BlindEditionNFT_Test.getEdition(id: self.editionID)!
        }

        /// Encode this edition object as a byte array.
        ///
        /// This can be used to hash the edition membership and verify its integrity.
        ///
        pub fun encode(): [UInt8] {
            return self.editionSalt
                .concat(self.editionID.toBigEndianBytes())
                .concat(self.editionSerial.toBigEndianBytes())
        }

        pub fun hash(): [UInt8] {
            return HashAlgorithm.SHA3_256.hash(self.encode())
        }
    }

    access(self) let editionMembers: {UInt64: EditionMember}

    pub fun getEditionMember(id: UInt64): EditionMember? {
        return BlindEditionNFT_Test.editionMembers[id]
    }

    pub resource NFT: NonFungibleToken.INFT, MetadataViews.Resolver {

        pub let id: UInt64

        /// A hash of the NFT's edition membership.
        ///
        /// The edition hash is known at mint time and 
        /// is generated by hashing the edition ID and
        /// serial number for this NFT.
        ///
        pub let editionHash: [UInt8]

        init(editionHash: [UInt8]) {
            self.id = self.uuid
            self.editionHash = editionHash
        }

        /// Return the edition that this NFT belongs to.
        ///
        /// This function returns nil if the edition membership has
        /// not yet been revealed.
        ///
        pub fun getEdition(): EditionMember? {
            return BlindEditionNFT_Test.editionMembers[self.id]
        }

        pub fun getViews(): [Type] {
            if self.getEdition() != nil {
                return [
                    Type<MetadataViews.NFTView>(),
                    Type<MetadataViews.Display>(),
                    Type<MetadataViews.ExternalURL>(),
                    Type<MetadataViews.NFTCollectionDisplay>(),
                    Type<MetadataViews.NFTCollectionData>(),
                    Type<MetadataViews.Royalties>()
                ]
            }

            return [
                Type<MetadataViews.ExternalURL>(),
                Type<MetadataViews.NFTCollectionDisplay>(),
                Type<MetadataViews.NFTCollectionData>(),
                Type<MetadataViews.Royalties>(),
                Type<MetadataViews.Display>(),
                Type<FreshmintMetadataViews.BlindNFT>()
            ]
        }

        pub fun resolveView(_ view: Type): AnyStruct? {
            if let editionMember = self.getEdition() {
                let edition = editionMember.getData()

                switch view {
                    case Type<MetadataViews.NFTView>():
                        return self.resolveNFTView(edition.metadata)
                    case Type<MetadataViews.Display>():
                        return self.resolveDisplay(edition.metadata)
                    case Type<MetadataViews.ExternalURL>():
                        return self.resolveExternalURL()
                    case Type<MetadataViews.NFTCollectionDisplay>():
                        return self.resolveNFTCollectionDisplay()
                    case Type<MetadataViews.NFTCollectionData>():
                        return self.resolveNFTCollectionData()
                    case Type<MetadataViews.Royalties>():
                        return self.resolveRoyalties()
                }

                return nil
            }

            switch view {
                case Type<MetadataViews.Display>():
                    return MetadataViews.Display(
                        name: \\"BlindEditionNFT_Test\\",
                        description: \\"This NFT is not yet revealed.\\",
                        thumbnail: MetadataViews.IPFSFile(
                            cid: BlindEditionNFT_Test.placeholderImage, 
                            path: nil
                        )
                    )
                case Type<FreshmintMetadataViews.BlindNFT>():
                    return FreshmintMetadataViews.BlindNFT(metadataHash: self.editionHash)
                case Type<MetadataViews.ExternalURL>():
                    return self.resolveExternalURL()
                case Type<MetadataViews.NFTCollectionDisplay>():
                    return self.resolveNFTCollectionDisplay()
                case Type<MetadataViews.NFTCollectionData>():
                    return self.resolveNFTCollectionData()
                case Type<MetadataViews.Royalties>():
                    return self.resolveRoyalties()
            }

            return nil
        }

        pub fun resolveNFTView(_ metadata: Metadata): MetadataViews.NFTView {
            return MetadataViews.NFTView(
                id: self.id,
                uuid: self.uuid,
                display: self.resolveDisplay(metadata),
                externalURL: self.resolveExternalURL(),
                collectionData: self.resolveNFTCollectionData(),
                collectionDisplay: self.resolveNFTCollectionDisplay(),
                royalties : self.resolveRoyalties(),
                traits: nil
            )
        }
        
        pub fun resolveDisplay(_ metadata: Metadata): MetadataViews.Display {
            return MetadataViews.Display(
                name: metadata.name,
                description: metadata.description,
                thumbnail: MetadataViews.IPFSFile(cid: metadata.thumbnail, path: nil)
            )
        }
        
        pub fun resolveExternalURL(): MetadataViews.ExternalURL {
            return MetadataViews.ExternalURL(\\"http://foo.com/\\".concat(self.id.toString()))
        }
        
        pub fun resolveNFTCollectionDisplay(): MetadataViews.NFTCollectionDisplay {
            let media = MetadataViews.Media(
                file: MetadataViews.IPFSFile(
                    cid: \\"bafkreicrfbblmaduqg2kmeqbymdifawex7rxqq2743mitmeia4zdybmmre\\", 
                    path: nil
                ),
                mediaType: \\"image/jpeg\\"
            )
        
            return MetadataViews.NFTCollectionDisplay(
                name: \\"My Collection\\",
                description: \\"This is my collection.\\",
                externalURL: MetadataViews.ExternalURL(\\"http://foo.com\\"),
                squareImage: media,
                bannerImage: media,
                socials: {}
            )
        }
        
        pub fun resolveNFTCollectionData(): MetadataViews.NFTCollectionData {
            return MetadataViews.NFTCollectionData(
                storagePath: BlindEditionNFT_Test.CollectionStoragePath,
                publicPath: BlindEditionNFT_Test.CollectionPublicPath,
                providerPath: BlindEditionNFT_Test.CollectionPrivatePath,
                publicCollection: Type<&BlindEditionNFT_Test.Collection{BlindEditionNFT_Test.BlindEditionNFT_TestCollectionPublic}>(),
                publicLinkedType: Type<&BlindEditionNFT_Test.Collection{BlindEditionNFT_Test.BlindEditionNFT_TestCollectionPublic, NonFungibleToken.CollectionPublic, NonFungibleToken.Receiver, MetadataViews.ResolverCollection}>(),
                providerLinkedType: Type<&BlindEditionNFT_Test.Collection{BlindEditionNFT_Test.BlindEditionNFT_TestCollectionPublic, NonFungibleToken.CollectionPublic, NonFungibleToken.Provider, MetadataViews.ResolverCollection}>(),
                createEmptyCollectionFunction: (fun (): @NonFungibleToken.Collection {
                    return <-BlindEditionNFT_Test.createEmptyCollection()
                })
            )
        }
        
        pub fun resolveRoyalties(): MetadataViews.Royalties {
            return MetadataViews.Royalties(BlindEditionNFT_Test.royalties)
        }
        
        destroy() {
            BlindEditionNFT_Test.totalSupply = BlindEditionNFT_Test.totalSupply - (1 as UInt64)

            emit Burned(id: self.id)
        }
    }

    pub resource interface BlindEditionNFT_TestCollectionPublic {
        pub fun deposit(token: @NonFungibleToken.NFT)
        pub fun getIDs(): [UInt64]
        pub fun borrowNFT(id: UInt64): &NonFungibleToken.NFT
        pub fun borrowBlindEditionNFT_Test(id: UInt64): &BlindEditionNFT_Test.NFT? {
            post {
                (result == nil) || (result?.id == id):
                    \\"Cannot borrow BlindEditionNFT_Test reference: The ID of the returned reference is incorrect\\"
            }
        }
    }

    pub resource Collection: BlindEditionNFT_TestCollectionPublic, NonFungibleToken.Provider, NonFungibleToken.Receiver, NonFungibleToken.CollectionPublic, MetadataViews.ResolverCollection {
        
        /// A dictionary of all NFTs in this collection indexed by ID.
        ///
        pub var ownedNFTs: @{UInt64: NonFungibleToken.NFT}

        init () {
            self.ownedNFTs <- {}
        }

        /// Remove an NFT from the collection and move it to the caller.
        ///
        pub fun withdraw(withdrawID: UInt64): @NonFungibleToken.NFT {
            let token <- self.ownedNFTs.remove(key: withdrawID) 
                ?? panic(\\"Requested NFT to withdraw does not exist in this collection\\")

            emit Withdraw(id: token.id, from: self.owner?.address)

            return <- token
        }

        /// Deposit an NFT into this collection.
        ///
        pub fun deposit(token: @NonFungibleToken.NFT) {
            let token <- token as! @BlindEditionNFT_Test.NFT

            let id: UInt64 = token.id

            // add the new token to the dictionary which removes the old one
            let oldToken <- self.ownedNFTs[id] <- token

            emit Deposit(id: id, to: self.owner?.address)

            destroy oldToken
        }

        /// Return an array of the NFT IDs in this collection.
        ///
        pub fun getIDs(): [UInt64] {
            return self.ownedNFTs.keys
        }

        /// Return a reference to an NFT in this collection.
        ///
        /// This function panics if the NFT does not exist in this collection.
        ///
        pub fun borrowNFT(id: UInt64): &NonFungibleToken.NFT {
            return (&self.ownedNFTs[id] as &NonFungibleToken.NFT?)!
        }

        /// Return a reference to an NFT in this collection
        /// typed as BlindEditionNFT_Test.NFT.
        ///
        /// This function returns nil if the NFT does not exist in this collection.
        ///
        pub fun borrowBlindEditionNFT_Test(id: UInt64): &BlindEditionNFT_Test.NFT? {
            if self.ownedNFTs[id] != nil {
                let ref = (&self.ownedNFTs[id] as auth &NonFungibleToken.NFT?)!
                return ref as! &BlindEditionNFT_Test.NFT
            }

            return nil
        }

        /// Return a reference to an NFT in this collection
        /// typed as MetadataViews.Resolver.
        ///
        /// This function panics if the NFT does not exist in this collection.
        ///
        pub fun borrowViewResolver(id: UInt64): &AnyResource{MetadataViews.Resolver} {
            let nft = (&self.ownedNFTs[id] as auth &NonFungibleToken.NFT?)!
            let nftRef = nft as! &BlindEditionNFT_Test.NFT
            return nftRef as &AnyResource{MetadataViews.Resolver}
        }

        destroy() {
            destroy self.ownedNFTs
        }
    }

    /// Return a new empty collection.
    ///
    pub fun createEmptyCollection(): @NonFungibleToken.Collection {
        return <- create Collection()
    }

    /// The administrator resource used to create editions,
    /// mint and reveal NFTs.
    ///
    pub resource Admin {

        /// Create a new NFT edition.
        ///
        /// This function does not mint any NFTs. It only creates the
        /// edition data that will later be associated with minted NFTs.
        ///
        pub fun createEdition(
            size: UInt,
            name: String,
            description: String,
            thumbnail: String,
        ): UInt64 {
            let metadata = Metadata(
                name: name,
                description: description,
                thumbnail: thumbnail,
            )

            let edition = Edition(
                id: BlindEditionNFT_Test.totalEditions,
                size: size,
                metadata: metadata
            )

            BlindEditionNFT_Test.editions[edition.id] = edition

            emit EditionCreated(edition: edition)

            BlindEditionNFT_Test.totalEditions = BlindEditionNFT_Test.totalEditions + (1 as UInt64)

            return edition.id
        }

        /// Mint a new NFT.
        ///
        /// To mint a blind edition NFT, specify its edition hash
        /// that can later be used to verify the revealed NFT's 
        /// edition ID and serial number.
        ///
        pub fun mintNFT(editionHash: [UInt8]): @BlindEditionNFT_Test.NFT {
            let nft <- create BlindEditionNFT_Test.NFT(editionHash: editionHash)

            emit Minted(id: nft.id, editionHash: editionHash)

            BlindEditionNFT_Test.totalSupply = BlindEditionNFT_Test.totalSupply + (1 as UInt64)

            return <- nft
        }

        /// Reveal a minted NFT.
        ///
        /// To reveal an NFT, publish its edition ID, serial number
        /// and unique salt value.
        ///
        pub fun revealNFT(
            id: UInt64,
            editionID: UInt64,
            editionSerial: UInt64,
            editionSalt: [UInt8]
        ) {
            pre {
                BlindEditionNFT_Test.editionMembers[id] == nil : \\"NFT has already been revealed\\"
            }

            let editionMember = EditionMember(
                editionID: editionID,
                editionSerial: editionSerial,
                editionSalt: editionSalt,
            )

            BlindEditionNFT_Test.editionMembers[id] = editionMember

            emit Revealed(
                id: id,
                editionID: editionID,
                editionSerial: editionSerial
            )
        }

        /// Set the royalty recipients for this contract.
        ///
        /// This function updates the royalty recipients for all NFTs
        /// minted by this contract.
        ///
        pub fun setRoyalties(_ royalties: [MetadataViews.Royalty]) {
            BlindEditionNFT_Test.royalties = royalties
        }
    }

    /// Return a public path that is scoped to this contract.
    ///
    pub fun getPublicPath(suffix: String): PublicPath {
        return PublicPath(identifier: \\"BlindEditionNFT_Test_\\".concat(suffix))!
    }

    /// Return a private path that is scoped to this contract.
    ///
    pub fun getPrivatePath(suffix: String): PrivatePath {
        return PrivatePath(identifier: \\"BlindEditionNFT_Test_\\".concat(suffix))!
    }

    /// Return a storage path that is scoped to this contract.
    ///
    pub fun getStoragePath(suffix: String): StoragePath {
        return StoragePath(identifier: \\"BlindEditionNFT_Test_\\".concat(suffix))!
    }

    priv fun initAdmin(admin: AuthAccount) {
        // Create an empty collection and save it to storage
        let collection <- BlindEditionNFT_Test.createEmptyCollection()

        admin.save(<- collection, to: BlindEditionNFT_Test.CollectionStoragePath)

        admin.link<&BlindEditionNFT_Test.Collection>(BlindEditionNFT_Test.CollectionPrivatePath, target: BlindEditionNFT_Test.CollectionStoragePath)

        admin.link<&BlindEditionNFT_Test.Collection{NonFungibleToken.CollectionPublic, BlindEditionNFT_Test.BlindEditionNFT_TestCollectionPublic, MetadataViews.ResolverCollection}>(BlindEditionNFT_Test.CollectionPublicPath, target: BlindEditionNFT_Test.CollectionStoragePath)
        
        // Create an admin resource and save it to storage
        let adminResource <- create Admin()

        admin.save(<- adminResource, to: self.AdminStoragePath)
    }

    init(admin: AuthAccount, placeholderImage: String) {

        self.version = \\"0.0.32\\"

        self.CollectionPublicPath = BlindEditionNFT_Test.getPublicPath(suffix: \\"Collection\\")
        self.CollectionStoragePath = BlindEditionNFT_Test.getStoragePath(suffix: \\"Collection\\")
        self.CollectionPrivatePath = BlindEditionNFT_Test.getPrivatePath(suffix: \\"Collection\\")

        self.AdminStoragePath = BlindEditionNFT_Test.getStoragePath(suffix: \\"Admin\\")

        self.placeholderImage = placeholderImage

        self.royalties = []

        self.totalSupply = 0
        self.totalEditions = 0

        self.editions = {}
        self.editionMembers = {}
        
        self.initAdmin(admin: admin)

        emit ContractInitialized()
    }
}
"
`;
